File: frontend/src/api/mysqlDatasource.ts
import type { ApiResponse } from './schemaManagement'
import { http } from './http'
import { currentUserId } from './llmConfig'

export interface MysqlDatasource {
  id: string
  name: string
  description: string
  host: string
  port: number
  defaultDatabase: string
  username: string
  owner: string
  isDefault: boolean
  status: string
  hasPassword: boolean
  passwordMasked: string
  createdAt: string
  updatedAt: string
}

export interface MysqlDatasourceInput {
  name: string
  description?: string
  host: string
  port?: number
  defaultDatabase?: string
  username: string
  password?: string
  owner?: string
  isDefault?: boolean
  status?: string
}

export interface MysqlDatasourceUpdateInput {
  name?: string
  description?: string
  host?: string
  port?: number
  defaultDatabase?: string
  username?: string
  password?: string
  owner?: string
  isDefault?: boolean
  status?: string
}

export interface TestConnectionResult {
  ok: boolean
  latencyMs: number | null
  error: string | null
}

const PREFIX = '/v1/mysql-datasources'

function headers(userId: string) {
  return { 'X-User-Id': userId }
}

function unwrap<T>(response: ApiResponse<T>): T {
  if (!response.success || response.code !== 200) {
    throw new Error(response.msg || `MySQL 数据源接口请求失败：${response.code}`)
  }
  return response.data
}

function asApiPromise<T>(request: unknown): Promise<ApiResponse<T>> {
  return request as Promise<ApiResponse<T>>
}

export async function listMysqlDatasources(userId = currentUserId()): Promise<MysqlDatasource[]> {
  return unwrap(await asApiPromise<MysqlDatasource[]>(http.get(PREFIX, { headers: headers(userId) })))
}

export async function getMysqlDatasource(id: string, userId = currentUserId()): Promise<MysqlDatasource> {
  return unwrap(await asApiPromise<MysqlDatasource>(http.get(`${PREFIX}/${id}`, { headers: headers(userId) })))
}

export async function createMysqlDatasource(
  payload: MysqlDatasourceInput,
  userId = currentUserId(),
): Promise<MysqlDatasource> {
  return unwrap(
    await asApiPromise<MysqlDatasource>(http.post(PREFIX, payload, { headers: headers(userId) })),
  )
}

export async function updateMysqlDatasource(
  id: string,
  payload: MysqlDatasourceUpdateInput,
  userId = currentUserId(),
): Promise<MysqlDatasource> {
  return unwrap(
    await asApiPromise<MysqlDatasource>(http.put(`${PREFIX}/${id}`, payload, { headers: headers(userId) })),
  )
}

export async function deleteMysqlDatasource(id: string, userId = currentUserId()): Promise<void> {
  await unwrap(
    await asApiPromise<{ deleted: boolean }>(http.delete(`${PREFIX}/${id}`, { headers: headers(userId) })),
  )
}

export async function setDefaultMysqlDatasource(id: string, userId = currentUserId()): Promise<MysqlDatasource> {
  return unwrap(
    await asApiPromise<MysqlDatasource>(
      http.post(`${PREFIX}/${id}/set-default`, {}, { headers: headers(userId) }),
    ),
  )
}

export async function testMysqlDatasource(id: string, userId = currentUserId()): Promise<TestConnectionResult> {
  return unwrap(
    await asApiPromise<TestConnectionResult>(
      http.post(`${PREFIX}/${id}/test`, {}, { headers: headers(userId) }),
    ),
  )
}

export async function listMysqlDatabases(id: string, userId = currentUserId()): Promise<string[]> {
  return unwrap(
    await asApiPromise<{ items: string[] }>(http.get(`${PREFIX}/${id}/databases`, { headers: headers(userId) })),
  ).items
}

File: backend/biz/schemas/mysql_datasource.py
"""平台 MySQL 数据源接口请求模型。"""

from __future__ import annotations

from pydantic import Field

from biz.schemas.schema_management import CamelModel


class MysqlDatasourceCreate(CamelModel):
    name: str = Field(min_length=1, max_length=128)
    description: str = Field(default="", max_length=500)
    host: str = Field(min_length=1, max_length=256)
    port: int = Field(default=3306, ge=1, le=65535)
    default_database: str = Field(default="", max_length=128)
    username: str = Field(min_length=1, max_length=128)
    password: str = Field(default="", max_length=256)
    owner: str = Field(default="", max_length=128)
    is_default: bool = False
    status: str = Field(default="正常")


class MysqlDatasourceUpdate(CamelModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=500)
    host: str | None = Field(default=None, min_length=1, max_length=256)
    port: int | None = Field(default=None, ge=1, le=65535)
    default_database: str | None = Field(default=None, max_length=128)
    username: str | None = Field(default=None, min_length=1, max_length=128)
    password: str | None = Field(default=None, max_length=256)
    owner: str | None = Field(default=None, max_length=128)
    is_default: bool | None = None
    status: str | None = Field(default=None, max_length=32)
